apis/v1alpha2/clusternetworkpolicy_types.go

```diff
@@ -0,0 +1,479 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// All fields in this package are required unless Explicitly marked optional
+// +kubebuilder:validation:Required
+package v1alpha2
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterNetworkPolicy is a cluster-wide network policy resource.
+//
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:shortName=cnp,scope=Cluster
+// +kubebuilder:printcolumn:name="Tier",type=string,JSONPath=".spec.tier"
+// +kubebuilder:printcolumn:name="Priority",type=string,JSONPath=".spec.priority"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterNetworkPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec defines the desired behavior of ClusterNetworkPolicy.
+	Spec ClusterNetworkPolicySpec `json:"spec"`
+
+	// Status is the status to be reported by the implementation.
+	// +optional
+	Status ClusterNetworkPolicyStatus `json:"status,omitempty"`
+}
+
+// ClusterNetworkPolicyStatus defines the observed state of
+// ClusterNetworkPolicy.
+type ClusterNetworkPolicyStatus struct {
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+// ClusterNetworkPolicySpec defines the desired state of ClusterNetworkPolicy.
+type ClusterNetworkPolicySpec struct {
+	// Tier is used as the top-level grouping for network policy prioritization.
+	//
+	// Policy tiers are evaluated in the following order:
+	// * Admin tier
+	// * NetworkPolicy tier
+	// * Baseline tier
+	//
+	// ClusterNetworkPolicy can use 2 of these tiers: Admin and Baseline.
+	//
+	// The Admin tier takes precedence over all other policies. Policies
+	// defined in this tier are used to set cluster-wide security rules
+	// that cannot be overridden in the other tiers.
+	//
+	// NetworkPolicy tier is the tier for the namespaced v1.NetworkPolicy.
+	// These policies are intended for the application developer to describe
+	// the security policy associated with their deployments inside their
+	// namespace.
+	//
+	// Baseline tier is a cluster-wide policy that can be overridden by the
+	// v1.NetworkPolicy. It will only be evaluated for Pods not selected by a
+	// namespaced v1.NetworkPolicy.
+	//
+	// If a given connection wasn't allowed or denied by any of the tiers,
+	// the default kubernetes policy is applied, which says that
+	// all pods can communicate with each other.
+	Tier Tier `json:"tier"`
```

Comment:
I think we haven't discussed this explicitly in the NPEP, but I am making both tier and priority fields required, so no default values. The main difference is that BANP will now have to set some priority, while before there was only 1 BANP allowed

Reply:
that is up for discussion - my thinking is that considering we are squashing 2 APIs into 1, it could be easy for people to upgrade to the wrong tier if we have any defaults (also keeping in mind that we wanted to have a very explicit API)